File: SortingAlgorithms/insertion_sort.cpp
#include<iostream>
#include<vector>
using namespace std;

// Time complexity: O(n^2), Space complexity: O(1)

// In-place algorithm, stable sorting algorithm
class InsertionSort {
public:
    void insSort(vector<int>& arr) {
        int n = arr.size();
        for (int i = 1; i < n; i++) {
            int val = arr[i];
            int hole = i; // Everything to the left of hole is considered sorted, insert val in the correct
            // position in subarr[0:i - 1]

            while(hole > 0 && arr[hole - 1] > val) {
                arr[hole] = arr[hole - 1];
                hole--;
            }
            arr[hole] = val;
        }

        return;
    }
};

int main() {
    vector<int> arr = {3, 4, 1, 9, -3, 9};

    cout << "Array before sorting" << endl;
    for (auto &x : arr) {
        cout << x << " ";
    }
    cout << endl;

    InsertionSort iSort;

    iSort.insSort(arr);

    cout << "Array after sorting" << endl;
    for (auto &x : arr) {
        cout << x << " ";
    }
    cout << endl;

    return 0;
}

File: SortingAlgorithms/insertion_sort.py
class InsertionSort:
    def insertion_sort(self, arr):

        n = len(arr)

        # Assuming that arr[0, ..., j - 1] is sorted.
        for j in range(1, n):
            val = arr[j]
            hole = j 
            while(hole > 0 and arr[hole - 1] > val):
                arr[hole] = arr[hole - 1]
                hole = hole - 1        
            arr[hole] = val

        return arr
    
def main():
    # Time complexity: O(n^2)
    arr = [9, 8, 7, 5, 6]

    insertObj = InsertionSort()
    arr_sort = insertObj.insertion_sort(arr)

    print("Array after sorting:\n")
    for elem in arr_sort:
        print(elem)

if __name__ == "__main__":
    main()